src/algebra/poly/authenticated_poly.rs

```diff
@@ -288,39 +288,117 @@ impl<C: CurveGroup> Mul<&AuthenticatedDensePoly<C>> for &AuthenticatedDensePoly<
+
+impl<C: CurveGroup> Mul<&ScalarResult<C>> for &AuthenticatedDensePoly<C> {
+    type Output = AuthenticatedDensePoly<C>;
+
+    fn mul(self, rhs: &ScalarResult<C>) -> Self::Output {
+        let new_coeffs = self.coeffs.iter().map(|coeff| coeff * rhs).collect_vec();
+        AuthenticatedDensePoly::from_coeffs(new_coeffs)
+    }
+}
+impl_borrow_variants!(AuthenticatedDensePoly<C>, Mul, mul, *, ScalarResult<C>, C: CurveGroup);
+impl_commutative!(AuthenticatedDensePoly<C>, Mul, mul, *, ScalarResult<C>, C: CurveGroup);
+
+impl<C: CurveGroup> Mul<&AuthenticatedScalarResult<C>> for &AuthenticatedDensePoly<C> {
+    type Output = AuthenticatedDensePoly<C>;
+
+    fn mul(self, rhs: &AuthenticatedScalarResult<C>) -> Self::Output {
+        let new_coeffs = self.coeffs.iter().map(|coeff| coeff * rhs).collect_vec();
+        AuthenticatedDensePoly::from_coeffs(new_coeffs)
+    }
+}
+impl_borrow_variants!(AuthenticatedDensePoly<C>, Mul, mul, *, AuthenticatedScalarResult<C>, C: CurveGroup);
+impl_commutative!(AuthenticatedDensePoly<C>, Mul, mul, *, AuthenticatedScalarResult<C>, C: CurveGroup);
+
+// --- Division --- //
+/// Given a public divisor b(x) and shared dividend a(x) = a_1(x) + a_2(x) for party shares a_1, a_2
+/// We can divide each share locally to obtain a secret sharing of \floor{a(x) / b(x)}
+///
+/// To see this, consider that a_1(x) = q_1(x)b(x) + r_1(x) and a_2(x) = q_2(x)b(x) + r_2(x) where:
+///     - deg(q_1) = deg(a_1) - deg(b)
+///     - deg(q_2) = deg(a_2) - deg(b)
+///     - deg(r_1) < deg(b)
+///     - deg(r_2) < deg(b)
+/// The floor division operator for a(x), b(x) returns q(x) such that there exists r(x): deg(r) < deg(b)
+/// where a(x) = q(x)b(x) + r(x)
+/// Note that a_1(x) + a_2(x) = (q_1(x) + q_2(x))b(x) + r_1(x) + r_2(x), where of course
+/// deg(r_1 + r_2) < deg(b), so \floor{a(x) / b(x)} = q_1(x) + q_2(x); making q_1, q_2 additive
+/// secret shares of the result as desired
+impl<C: CurveGroup> Div<&DensePolynomialResult<C>> for &AuthenticatedDensePoly<C> {
+    type Output = AuthenticatedDensePoly<C>;
+
+    fn div(self, rhs: &DensePolynomialResult<C>) -> Self::Output {
+        // We cannot break early if the remainder is exhausted because this will cause the gate
+        // sequencing to differ between parties in the MPC. Instead we execute the whole computation on
+        // both ends of the MPC
+        assert!(!rhs.coeffs.is_empty(), "cannot divide by zero polynomial");
+        let fabric = self.coeffs[0].fabric();
+
+        let quotient_degree = self.degree().saturating_sub(rhs.degree());
+        if quotient_degree == 0 {
+            return AuthenticatedDensePoly::zero(fabric);
+        }
+
+        let mut remainder = self.clone();
+        let mut quotient_coeffs = fabric.ones_authenticated(quotient_degree + 1);
+
+        let divisor_leading_inverse = rhs.coeffs.last().unwrap().inverse();
+        for deg in (0..=quotient_degree).rev() {
```

Comment:
in the implementations i'm seeing (pseudocode for long division / euclidean division on wikipedia and the Arkworks `divide_with_q_and_r`), this is structured as a `while remainder != 0 && deg(remainder) >= deg(divisor)`, because it seems the remainder's degree can decrease by more than 1 in each round. E.g. in Arkworks, each while loop iteration ends with a "`pop the leading coeff of the remainder until it's no longer zero`".

I suppose here we assume one degree per loop due to the secret-shared nature of the remainder coefficients? I.e. if I have a zero for the current leading coefficient of the remainder, my MPC counterparty may not?

That said, is it also guaranteed that the remainder's degree decreases by _at least_ 1 in each iteration?

Reply:
Technically no it is not guaranteed, you can imagine the remainder having a leading coefficient of `0` on some iteration. That said, we know that after `quotient_degree + 1` iterations (as above), the remainder is of degree less than the divisor, which is the only invariant a floor div needs.

Reply:
ah yup i see now how the remainder's leading coefficient gets cleared out on every iteration, ack